### PR TITLE
Optimize 10b 2b sdpa moe

### PR DESCRIPTION
**Pull request by nscottnichols**
Added some optimizations for the MoE path, wanted to PR before our forks drift too far apart

## Summary by Sourcery

Optimize Mixture-of-Experts routing, dispatch, and expert execution paths for better load balancing and no-grad performance while preserving numerical behavior.

New Features:
- Add optional force-load-balance MoE token dispatcher path with equal-split all-to-all routing and debug counters controllable via environment variables.
- Introduce batched no-grad expert execution and cached weight transformations for uniform per-expert token counts.
- Provide in-place deterministic scatter_add for forward-only paths and integrate it into MoE combine steps.

Enhancements:
- Extend token dispatch metadata and routing to reuse precomputed per-expert token counts across router, dispatcher, and experts, avoiding redundant histogram computation.
- Optimize all-to-all token dispatch/combine with equal-padding fast paths and shape-aware permute/unpermute logic, including specialized handling for TorchAO dispatch.
- Improve MoE scoring paths by avoiding unnecessary dtype upcasts/ casts where possible and by differentiating bf16 scoring behavior for monitoring.
- Make expert FSDP sharding robust to PyTorch builds lacking per-parameter mesh APIs by falling back to a two-phase sharding strategy.

Tests:
- Add extensive unit tests for MoE fast-path counters, deterministic scatter_add in-place behavior, token dispatcher correctness, equal-split and force-load-balance routing logic, and equal-count permute/unpermute paths.

### CODE REVIEW

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -27,30 +27,46 @@ def _run_experts_for_loop(
...
+
+    if (
+        len(num_tokens_per_expert_list) > 0
+        and all(count == num_tokens_per_expert_list[0] for count in num_tokens_per_expert_list)
+        and num_tokens_per_expert_list[0] > 0
+        and not torch.is_grad_enabled()
+    ):
+        _record_moe_fastpath("batched_no_grad_experts")
+        tokens_per_expert = num_tokens_per_expert_list[0]
+        x_grouped = x.view(len(num_tokens_per_expert_list), tokens_per_expert, x.shape[-1])
```
[the remark this answers is not in the document]
Comment (nscottnichols) [reply torchtitan/models/common/moe.py]: 1. I don't think these are real problems
2. `x` is contiguous, isn't it?

File: torchtitan/models/common/token_dispatcher.py [hunk 2]
```diff
@@ -178,6 +275,165 @@ def __init__(self, config: Config):
...
+
+        Round-robin force-load-balanced routing with ``top_k == num_local_experts``
+        sends each EP rank the same strided token shard once per local expert.
+        Materializing that rank-major equal-split layout directly avoids first
+        gathering the unpadded expert-sorted buffer and then padding each rank
+        segment.
+        """
+        pieces = []
+        for rank, split in enumerate(input_splits):
+            tokens_for_rank = split // num_local_experts
```
[the remark this answers is not in the document]
Comment (nscottnichols) [reply torchtitan/models/common/token_dispatcher.py]: Good point

**Comment by nscottnichols** [comment]
@saforem2 do you want me to fix this sourcery junk or will you handle it? Some of them are good points.

**Comment by saforem2** [comment]
😂 @nscottnichols honestly whatever you think; like you said, some of them are good points but some of them are either nitpicky or incorrect

My only actual question would be, can we move these changes to be self-contained in the `experiments/ezpz/` directory somehow?

i.e. could we move the changes from maybe:

```bash
models/common/moe.py --> experiments/ezpz/moe/model.py
models/common/token_dispatcher.py --> experiments/ezpz/moe/token_dispatcher.py
ops/scatter_add.py --> experiments/ezpz/moe/token_dispatcher.py
```

I guess this can be a PITA if it continues to grow; but if there are useful changes outside of `experiments/ezpz/` we can maybe split them out and submit them as PRs to upstream.

Also, incase you haven't seen [this thread](https://github.com/saforem2/torchtitan/pull/10#issuecomment-4425963332) in #10, it looks like there are upstream changes that break things:

> upstream just landed https://github.com/pytorch/torchtitan/pull/3308 (b301dfa0, "Remove MoE expert for-loop fallback") which removes use_grouped_mm from GroupedExperts.Config entirely and inlines _grouped_mm as the only path. [...]

**Comment by nscottnichols** [comment]
>My only actual question would be, can we move these changes to be self-contained in the experiments/ezpz/ directory somehow?

Let me think about how this can be done. There are definitely some parts that can be moved.

**Comment by saforem2** [comment]
Expert-side pieces folded into #12 (commit `364fbedf`):

- Equal-counts no-grad bmm fast path inside the for-loop backend
- Version-keyed `w13` / `w2_t` caches on `EzpzGroupedExperts`
- `_record_moe_fastpath` counters (env-gated)
- The FSDP `try/except ImportError` two-phase fallback (commit `09a7c697`, layered on top of #9's HSDPMeshInfo fix)

**Deferred** to a follow-up: the token-dispatcher rewrite in `torchtitan/models/common/token_dispatcher.py` (`+570/-91`), `torchtitan/ops/scatter_add.py` in-place wrapper, and `experiments/ezpz/moe/model.py` `force_load_balance` flag-through. Keeping that all in `experiments/ezpz/` needs an `EzpzLocalTokenDispatcher` subclass which is a meaningful chunk on its own. Going to keep this open as the home for that follow-up.

File: torchtitan/models/common/moe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nscottnichols) [reply torchtitan/models/common/moe.py]: Addressed in 9175ec8. I made the layout contract explicit by checking the expected numel and switched the fast path from view(...) to reshape(...). Current behavior is unchanged for contiguous expert-major inputs, but this is safer if a future dispatch path changes layout.

File: torchtitan/models/common/token_dispatcher.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nscottnichols) [reply torchtitan/models/common/token_dispatcher.py]: Addressed in 9175ec8. Replaced the unused local variable with an explicit assert that each split is divisible by num_local_experts, which is the invariant this direct force-load-balanced layout relies on. Added a unit test for the rejection case.

File: torchtitan/ops/scatter_add.py [hunk 3]
```diff
@@ -20,6 +20,19 @@ def deterministic_scatter_add(
         torch.use_deterministic_algorithms(prev, warn_only=prev_warn_only)
```
[the remark this answers is not in the document]
Comment (nscottnichols) [reply torchtitan/ops/scatter_add.py:20]: I did not add a fake registration here because deterministic_scatter_add_ is a plain Python wrapper around out.scatter_add_, not a torch.library custom_op. The existing fake registration applies to the out-of-place custom op. If we later convert the in-place helper into a custom op, then it should get a matching fake implementation.

File: tests/unit_tests/test_expert_parallel.py [hunk 4]
```diff
@@ -5,11 +5,427 @@
...
 
 import unittest
+import os
 
 import torch
 import torch.nn.functional as F
+from torch import nn
+
+from torchtitan.models.common.linear import Linear
+from torchtitan.models.common.moe import TokenChoiceTopKRouter
```
[the remark this answers is not in the document]
Comment (nscottnichols) [reply tests/unit_tests/test_expert_parallel.py]: Addressed in 9175ec8. Added focused tests for numerical equivalence of the batched no-grad expert path and for w13 / w2_t cache hit, miss, and invalidation behavior through the fast-path counters.

**Comment by nscottnichols** [comment]
> Hey - I've found 5 issues, and left some high level feedback:
> 
> * In `_pad_to_equal_splits`, the `F.pad(piece, (0, 0, 0, equal_split_size - split))` call pads extra rows at the _front_ rather than the _end_ of each segment, which contradicts the new test expectations (and the intended layout of original tokens followed by padding); consider swapping the height pads to `(0, 0, equal_split_size - split, 0)` or using `torch.cat` with an explicit zeros tensor instead.
> * The `_shares_storage` helper only compares `untyped_storage().data_ptr()` and ignores storage offsets, so two views into the same underlying storage at different offsets will be treated as non-aliasing; this can allow `_scatter_add_forward_or_autograd` to perform in-place updates on tensors that still overlap `x`, potentially corrupting inputs—consider also checking `storage_offset()` or using `torch._unsafe_view`-based alias checks.


I checked this and did not keep a code change here. For a 2D tensor, F.pad(piece, (0, 0, 0, n)) pads rows at the end, which matches the existing unit test expectations. I also tried replacing it with explicit zero-row concatenation, but that regressed the benchmark from ~29.9 ms to 31.514 ms, so I reverted that experiment.

**Comment by nscottnichols** [comment]
Oh, I also didn't change the "shares storage" guard. The current guard is intentionally conservative for the known alias case we hit: shared_experts(x) returning x exactly. If we want broader partial-overlap detection for arbitrary views, that is worth a separate patch because it can affect the no-grad in-place scatter fast path.

**Comment by saforem2** [comment]
FYI — #13 landed the upstream resync onto `ezpz` (pulls pytorch/torchtitan#3308 which removed `use_grouped_mm` and `_run_experts_for_loop` from core). The replay introduced `EzpzGroupedExperts(GroupedExperts)` in `torchtitan/experiments/ezpz/moe/experts.py` with a vendored copy of `_run_experts_for_loop`.

When rebasing this PR onto merged `ezpz`, the expert-side optimizations (cached `w13` / `w2_t` + equal-counts no-grad bmm fast path + counters) should layer onto that vendored `_run_experts_for_loop` in `experts.py` rather than `models/common/moe.py`. The token-dispatcher rewrite is unaffected by #13. Your HSDP `try/except ImportError` import-fallback in `parallelize.py` is also unaffected and should rebase cleanly.

Withdrawing the prior #12 which had attempted to bundle these in — keeping #11 as the canonical home for the optimizations.

**Comment by saforem2** [comment]
Closing as superseded after auditing the implementation paths. The production-relevant dispatcher, equal-all-to-all, forced-routing, deterministic-combine, expert, and FSDP work is integrated and further hardened in the current line. The only absent residue is an old standalone benchmark harness; if still useful, it should be revalidated and proposed separately rather than reviving this broad stale branch.
